### PR TITLE
Fixing missing seeds affecting Electron EcalDriven Efficiency

### PR DESCRIPTION
**Pull request by Sam-Harper**
Dear All,

The efficiency of an electron to be ECAL driven has been low by 7-10% in 2017 data and this is one of the show stopping issues for the data re-reco. A reminder, electrons can be tracker driven (seeded by tracks) or ecal driven (seeded by pixel hits matched to the supercluster's trajectory). The tracker driven electrons were fine but there were issues with the ecal driven electrons

The issue has now been identified and was discussed today in the e/gamma meeting:
https://indico.cern.ch/event/604940/contributions/2686772/attachments/1506133/2347059/ecalDrivenFix.pdf

To summarise, E/gamma wants to have as input all sensible doublet combinations. This is done by pixelPairElectronSeeds but importantly it is cleaned for previously made pixel seeds, ie a cluster cant be considered for pixelPairElectronSeeds if it was used for a previous seed collection.  A final seed collection, newCombinedSeeds is made out of these seeds and all the previously made pixel seeds for which it was cleaned. This is presumably to reduce timing and is a sensible approach. Unfortunately for phase1 pixels, a rather key set of seeds, pixelPairSeeds, were removed from newCombinedSeeds. 

To summarise the summary:  [the chap guessed wrong](https://github.com/cms-sw/cmssw/blob/master/RecoTracker/IterativeTracking/python/ElectronSeeds_cff.py#L37)

I have added them back along with all other seeds which are involved in the cleaning. This covers most of the inefficiency but there remains a 0.5% effect compared to just making the doublets directly without cleaning. I will continue to have a look at this but things are very much improved already so I think its important to have into the release asap.

To do:
I want to re-reco all of the DoubleEG dataset of 299480 just to check there are no adverse effects from this (I only re-recoed the failing ecal driven electrons). I expect this to complete tomorrow, grid willing. 

Also this potentially impacts timing. What is the recommended benchmark job to evaluate this impact? 

Finally this wasnt supposed to be in the MC but given the fix, its weird it isnt there, to be investigated.

### CODE REVIEW

**Comment by slava77** [comment]
On 8/9/17 12:31 PM, Sam-Harper wrote:
> Also this potentially impacts timing. What is the recommended benchmark
> job to evaluate this impact?
> 

we usually use 10224.0 MC workfow after stripping off the validation&DQM
parts in it.

If you can instead get the numbers from 299480, which apparently has a
similar PU~30 as the MC,
it should be good enough.

File: RecoTracker/IterativeTracking/python/ElectronSeeds_cff.py [hunk 1]
```diff
@@ -246,7 +246,12 @@
     'pixelLessStepSeeds',
     'tripletElectronSeeds',
     'pixelPairElectronSeeds',
-    'stripPairElectronSeeds'
+    'stripPairElectronSeeds',
+    'lowPtTripletStepSeeds',
+    'lowPtQuadStepSeeds',
```
Comment (slava77) [inline RecoTracker/IterativeTracking/python/ElectronSeeds_cff.py]: do the low-pt seeds actually help?

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by makortel** [comment]
> Unfortunately for phase1 pixels, a rather key set of seeds, pixelPairSeeds, were removed from newCombinedSeeds.

At the time pixelPairStep was not part of phase1 tracking, so something had to be done (and I admit I didn't fully understand the logic in this file). Unfortunately it was not added here when we re-added the pixelPairStep (in #19051 that got in 9_2_2).

Taking a look now I would think that including `lowPtTripletStepSeeds` alone should give the improvement as those are the one whose clusters get masked out (the module is called `detachedTripletStepSeedClusterMask` but the input is `lowPtTripletStepSeeds` as Sam wrote down on slide 4).

If the point is (not sure but it looks like that) to mask the clusters used in seeds of pixel iterations targeting "high-pT prompt tracks" and then add the seeds of these iterations to `newCombinedSeeds` (*), then maybe the best pixel-only list (at least in case all seeds are too much) would be `initialStep, highPtTripletStep, pixelPairStep` (i.e. phase0 list + highPtTripletStep)? 

(*) and if this is indeed the case, I would like to get some explanatory comments on the top of the file to explain the logic

**Comment by Sam-Harper** [comment]
Thanks Matti. 

I have to admit, i dont really know what the point of all this. E/gamma wants all sensible doublet combinations to go into newCombinedSeeds and it looks to save CPU, it cleans previously made ones. Which mostly works. 

The lowPtTripletsStepSeeds does indeed give a large improvement as you suggested (well I only tested removeing both lowPtTripletStepSeeds and lowPtQuadStepSeeds). The detached seeds have a very small effect. 

You can see it from this plot:
https://sharper.web.cern.ch/sharper/cms/heep/2017/Aug9th/ecalDrivenEffVsSeedComp.gif

And given your explanation of the issue, I think we now know why we didnt see it in the MC (pre 922 release) and its probably there  now :(

In fact I can see it in this plot:
http://cms-egamma.web.cern.ch/cms-egamma/validation/Electrons/Releases/9_2_2_DQM_std/GedvsGed_9_2_1_DQM_std/Fullgedvsged_RelValSingleElectronPt1000_gedGsfE_Startup/gifs/h_ele_provenance_barrel.gif

File: RecoTracker/IterativeTracking/python/ElectronSeeds_cff.py (hunk 1 above)
Comment (Sam-Harper) [inline RecoTracker/IterativeTracking/python/ElectronSeeds_cff.py]: so as Matti suggested and I then proved in the reply to him, the low pt seeds significantly help. The detached seeds dont help that much but I would prefer to have them unless timing forces me not to.

**Comment by Sam-Harper** [comment]
Dear All,

I have completed my timing tests:
It did it on step3 10224.0 MC 

step3Test --conditions auto:phase1_2017_realistic --pileup_input das:/RelValMinBias_13/CMSSW_9_2_3-92X_upgrade2017_realistic_v1_earlyBS2017-v1/GEN-SIM -n 10 --era Run2_2017 --eventcontent RECOSIM,MINIAODSIM --runUnscheduled -s RAW2DIGI,L1Reco,RECO,EI,PAT --datatier GEN-SIM-RECO,MINIAODSIM --pileup AVE_35_BX_25ns --geometry DB:Extended --filein file:step2.root --fileout file:step3.root

with 4 threads running, nothing else on the machine for 200 events (from the previous step2 output). Please let me know if I should modify my test or run on more events. 

~/tools/timeDiffFromReport.sh defaultRecoTest2.log allSeedsRecoTest3.log 
Couldn't parse time report using CPU and wall-clock format: trying Wall-clock only
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
   -1.652563      -0.10%         9.39 ms/ev ->         0.89 ms/ev dt1DCosmicRecHits
   +0.854941      +0.09%         5.05 ms/ev ->        12.60 ms/ev siStripZeroSuppression
   +0.337662      +0.00%         0.54 ms/ev ->         0.76 ms/ev basicJetsForMetNoHF
   +0.221314      +0.03%        11.10 ms/ev ->        13.86 ms/ev ecalDrivenElectronSeeds
   -0.220901      -0.00%         1.12 ms/ev ->         0.90 ms/ev shiftedPatJetResUpNoHF
   +0.215043      +0.00%         0.88 ms/ev ->         1.10 ms/ev shiftedPatSmearedJetResUpNoHF
   -0.212594      -0.00%         1.10 ms/ev ->         0.89 ms/ev patSmearedJetsPuppi
   -0.179495      -0.03%        13.05 ms/ev ->        10.90 ms/ev initialStepSeedsPreSplitting
   +0.168264      +0.04%        16.39 ms/ev ->        19.40 ms/ev initialStepHitQuadrupletsPreSplitting
   -0.100302      -0.02%        18.25 ms/ev ->        16.51 ms/ev detachedTripletStepSeeds
   +0.097561      +0.01%         5.54 ms/ev ->         6.11 ms/ev gedGsfElectrons
   -0.095300      -0.03%        27.54 ms/ev ->        25.03 ms/ev pixelPairElectronSeeds
   -0.091277      -0.01%         5.56 ms/ev ->         5.07 ms/ev muonCSCDigis
   -0.079560      -0.01%        11.16 ms/ev ->        10.31 ms/ev lowPtQuadStepSeeds
   -0.075583      -0.01%         6.67 ms/ev ->         6.19 ms/ev highPtTripletStepSeeds
   -0.070692      -0.01%         7.03 ms/ev ->         6.55 ms/ev lowPtTripletStepSeeds
   +0.064615      +0.01%        15.20 ms/ev ->        16.22 ms/ev gedPhotons
   -0.062707      -0.01%        13.24 ms/ev ->        12.44 ms/ev initialStepSeeds
   +0.062573      +0.01%        10.39 ms/ev ->        11.06 ms/ev ctppsDiamondRecHits
   -0.058613      -0.05%        72.32 ms/ev ->        68.20 ms/ev pixelLessStepSeeds
  ---------- ------------     --------                  ----       ------------
Job total:  8.45768 s/ev ==> 8.41231 s/ev

I did this 3 times for current settings and new settings and while I saw fluctuations in other modules, I always got something like

   +0.221314      +0.03%        11.10 ms/ev ->        13.86 ms/ev ecalDrivenElectronSeeds

which is where I expect to see the change from this.

**Comment by slava77** [comment]
@Sam-Harper 
please also check the reco file size (hopefully the electron seeds do not overtake other large collections in size)

**Comment by Sam-Harper** [comment]
still based on previous setup, 200 events, ttbar. 

Summary, a small change cut barely noticeable 

with all seeds: 
recoElectronSeeds_electronMergedSeeds__RECO. 11454.4 1981.61
current seeds:
recoElectronSeeds_electronMergedSeeds__RECO. 9406.67 1840.82

didnt really impact the gsftracks

with all seeds: 
recoGsfTrackExtras_electronGsfTracks__RECO. 26601.9 18381.9
recoTrackExtras_electronGsfTracks__RECO. 6027.74 4176.73
recoGsfTracks_electronGsfTracks__RECO. 3186.01 1423.95
TrackingRecHitsOwned_electronGsfTracks__RECO. 7029.48 1043.12

current seeds:
recoGsfTrackExtras_electronGsfTracks__RECO. 26354.7 18211.5
recoTrackExtras_electronGsfTracks__RECO. 5968.49 4137.34
recoGsfTracks_electronGsfTracks__RECO. 3155.38 1410.68
TrackingRecHitsOwned_electronGsfTracks__RECO. 6953.88 1036.6

file sizes:
-rw-r--r-- 1 mjf21517 cms 585005444 Aug 11 16:13 recoTestAllSeeds.root
-rw-r--r-- 1 mjf21517 cms 584995648 Aug 11 16:13 recoTestAllSeeds2.root
-rw-r--r-- 1 mjf21517 cms  12477322 Aug 11 16:13 recoTestAllSeeds2_inMINIAOD.root
-rw-r--r-- 1 mjf21517 cms 585001538 Aug 11 16:13 recoTestAllSeeds3.root
-rw-r--r-- 1 mjf21517 cms  12468480 Aug 11 16:13 recoTestAllSeeds3_inMINIAOD.root
-rw-r--r-- 1 mjf21517 cms  12474822 Aug 11 16:13 recoTestAllSeeds_inMINIAOD.root
-rw-r--r-- 1 mjf21517 cms 584972464 Aug 11 16:03 recoTestStdSeeds.root
-rw-r--r-- 1 mjf21517 cms 584925605 Aug 11 16:09 recoTestStdSeeds2.root
-rw-r--r-- 1 mjf21517 cms  12465186 Aug 11 16:09 recoTestStdSeeds2_inMINIAOD.root
-rw-r--r-- 1 mjf21517 cms 584945263 Aug 11 16:09 recoTestStdSeeds3.root
-rw-r--r-- 1 mjf21517 cms  12463389 Aug 11 16:09 recoTestStdSeeds3_inMINIAOD.root
-rw-r--r-- 1 mjf21517 cms  12477855 Aug 11 16:03 recoTestStdSeeds_inMINIAOD.root

**Comment by slava77** [comment]
> Summary, a small change cut barely noticeable

thanks, the sizes look OK

**Comment by perrotta** [comment]
I've looked for possible side effects propagated to other high level objects in a similar 10224.0 MC, and I don't see anything worrying: there are a few more electrons in the event, as wanted, and PF and other quantities are touched in a apparently healthy way.

The number of electron seeds increases by some 15%, especially for 3 and 4 hits seeds:

all_tespr20106vsorig_ttbar13tevpu2017wf10224p0c_recoelectronseeds_electronmergedseeds__reco_obj_nhits

@Sam-Harper : are you done with the your planned re-reco all of the DoubleEG dataset of 299480, as anticipated in the description of this PR? As soon as you confirm that there is norhing worrying also there, I think we can sign off this PR

**Comment by Sam-Harper** [comment]
@perrotta. The re-reco went a bit more problematic than I had assumed (both technical and results). I (re-)discovered interesting things. For example this doesnt strictly increase efficiency. There are now events which fail due to track matching when before they would pass. This happens at the percent level. Expect a more detailed report tomorrow. I think I'll probably sign off on it though but this just keeps telling me that an e/gamma review of electron tracking is well well overdue.

**Comment by Sam-Harper** [comment]
Here is some plots on the matter https://sharper.web.cern.ch/sharper/cms/talks/validationOfTheElectronSeedFix.pdf

Short summary: by adding seeds, I actually cause an efficiency drop in the endcap if you dont care about ECAL driven. There is probably some background contamination in there but even if there isnt, this efficiency loss is at worst 0.5% (eta>1.0) and I think its worth it to fix the ECAL driven seeds. Note this loss is if you dont care about ecalDriven, if you do, the efficiency is of course much much better. I think its worth it. 

This is something that will be studied as part of our long term tracking studies. 

So I think we are good to go

**Comment by perrotta** [comment]
Thank you Sam.

Let me check if I understood your plots and explanations. If we re-reco with this PR we get:
- Efficiency for ECal driven electron largely improved
- Efficiency for Tracker driver electrons slightly reduced, mostly in the endcaps
- Efficiency for electrons as a whole (Ecal + Tracker driven) basically unchanged

This is my understanding, but I don't have a plot which confirms the third bullet in your slides.
Could you please either confirm or correct what above?

**Comment by Sam-Harper** [comment]
The 3rd plot is difficult to measure quickly but we assume as you say. 

 And the other two points you are correct. Again this new behavior is the intended one, before it is definitely a bug.

**Comment by perrotta** [comment]
Tested also on a high pt  QCD sample (200 events from QCD13TeVFlatPt15s3000in2017 wf10071p0).

Even there, no "strange" effects propagate to the electrons or higher level quantities.

As in the TTbar, the additional seeds with three and four hits are visible:

image

and the number of reconstructed electrons slighlty increases (from 114 to 116), as well as the gsf tracks (from 1483 to 1490):
image

ged photons reduce by three, see theta distributions:
image

**Comment by perrotta** [comment]
+1
- Results (https://github.com/cms-sw/cmssw/pull/20106#issuecomment-322138320) are in accordance with the EGamma POG expectations (https://github.com/cms-sw/cmssw/pull/20106#issuecomment-322163803)
- No unwanted side effects were observed in some TTbar (https://github.com/cms-sw/cmssw/pull/20106#issuecomment-321962242) and high pt QCD MC sample (https://github.com/cms-sw/cmssw/pull/20106#issuecomment-322481126)
- Timing (https://github.com/cms-sw/cmssw/pull/20106#issuecomment-321821111) and additional sizes (https://github.com/cms-sw/cmssw/pull/20106#issuecomment-321843653) can be considered acceptable

**Comment by Sam-Harper** [comment]
thanks @perrotta

**Comment by davidlange6** [comment]
merge

**Pull request merged**